### PR TITLE
lsif-typed: Add diagnostics

### PR DESCRIPTION
**Pull request by efritz**

### CODE REVIEW

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 1]
```diff
@@ -210,4 +210,13 @@ message Occurrence {
...
+  // (optional) Diagnostic information about this range.
+  repeated Diagnostic diagnostics = 6;
+}
+
+message Diagnostic {
+  int32 severity = 1;
+  string code = 2;
+  string message = 3;
+  string source = 4;
 }
```
Comment (efritz) [inline proto/lib/codeintel/lsif_typed/lsif.proto:222]: Please suggest fields/documentation (or existing documentation to steal).

**Review by olafurpg — APPROVED** [review]
I pushed a commit adding docstrings and moved diagnostics to become a field of `Document`.

File: proto/lib/codeintel/lsif_typed/lsif.proto (hunk 1 above)
Comment (tjdevries) [reply proto/lib/codeintel/lsif_typed/lsif.proto:240]: Other fields are:

```
tag: DiagnosticTag = {
    -- Unused or unnecessary code
    Unnecessary = 1;
    -- Deprecated or obsolete code
    Deprecated = 2;
  };

/**
 * Represents a related message and source code location for a diagnostic.
 * This should be used to point to code locations that cause or are related to
 * a diagnostics, e.g when duplicating a symbol in a scope.
 */
export interface DiagnosticRelatedInformation {
	/**
	 * The location of this related diagnostic information.
	 */
	location: Location;

	/**
	 * The message of this related diagnostic information.
	 */
	message: string;
}
```

The related information one is cool and I've seen it used by rust-analyzer in LSP to basically say "hey, you defined this here like this... but over here you use it like that".

**Comment by efritz** [comment]
@olafurpg Why nest diagnostics if nothing else is nested? I'm confused about the rule to use here - seems inconsistent.

File: proto/lib/codeintel/lsif_typed/lsif.proto (hunk 1 above)
Comment (efritz) [reply proto/lib/codeintel/lsif_typed/lsif.proto:240]: I think we've historically had custom extensions to LSP as well. (cc @beyang I know you've touched similar code, could you link us to any historic decision data we have?)

**Comment by olafurpg** [comment]
I'm OK with making diagnostics a part of `Occurrence`. My reasoning to separate was mostly:

- that's how it's done in SemanticDB
- diagnostic ranges are commonly multiline while  symbol occurrences and highlighting ranges are more commonly single-line
- I don't know what we use the diagnostics for

Feel free to move the diagnostic back under `Occurrence`, LGTM 👍 either way

File: proto/lib/codeintel/lsif_typed/lsif.proto (hunk 1 above)
Comment (efritz) [reply proto/lib/codeintel/lsif_typed/lsif.proto:295]: Added tags.

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 2]
```diff
@@ -261,4 +261,35 @@ message Occurrence {
   repeated string override_documentation = 4;
   // (optional) What syntax highlighting class should be used for this range?
   SyntaxKind syntax_kind = 5;
+  // Diagnostics that have been reported for this specific range.
+  repeated Diagnostic diagnostics = 6;
```
Comment (efritz) [inline proto/lib/codeintel/lsif_typed/lsif.proto:265]: @olafurpg I'd like to discuss more your intention of splitting this out of occurrence.

Comment (olafurpg) [reply proto/lib/codeintel/lsif_typed/lsif.proto:265]: I don't have a strong opinion whether this is included with occurrence or if we use `Document.diagnostics`. The `Occurrence` range is more typically a single-line while diagnostics are frequently multi-line so I find it a bit more natural to keep them separate, but it's not a strong argument.

Comment (efritz) [reply proto/lib/codeintel/lsif_typed/lsif.proto:265]: > The Occurrence range is more typically a single-line while diagnostics are frequently multi-line

Where are you getting this assumption? According to our protocol, an occurrence just holds *some data* about *some sequence of characters* in a text document. Why is there more meaning prescribed behind it in your argument?

I can think of plenty of diagnostics that aren't necessarily multi-line (use of deprecated field or method).

Comment (efritz) [reply proto/lib/codeintel/lsif_typed/lsif.proto:265]: Could you explain what the original assumptions were so I can convince myself this is correct?

Comment (olafurpg) [reply proto/lib/codeintel/lsif_typed/lsif.proto:265]: It was mostly the three points in this comment here https://github.com/sourcegraph/sourcegraph/pull/29886#issuecomment-1016681071

**Pull request merged**